syntax = "proto3";

package flashreduce_proto;

service Session {
  rpc RdmaSession (RdmaSessionRequest) returns (RdmaSessionResponse) {}
}

service Sync {
  rpc Barrier(BarrierRequest) returns (BarrierResponse) {}
  rpc Broadcast(BroadcastRequest) returns (BroadcastResponse) {}
}


message RdmaSessionRequest {
  uint32 session_id = 1;
  uint32 rank = 2;
  uint32 num_workers = 3;
  uint32 root = 4;
  uint64 mac = 5;
  uint32 ipv4 = 6;
  uint32 rkey = 7;
  uint64 raddr = 8;
  uint32 qpn = 9;
  uint32 psn = 10;
  uint64 gid_subnet = 11;
  uint64 gid_iface = 12;
  uint32 lid = 13;
}

message RdmaSessionResponse {
  uint32 rkey = 2;
  uint64 raddr = 3;
  uint32 qpn = 4;
  uint32 psn = 5;
  uint64 gid_subnet = 13;
  uint64 gid_iface = 14;
  uint32 lid = 15;
}

message BarrierRequest {
  uint32 num_workers = 1;
}

message BarrierResponse {
}

message BroadcastRequest {
  uint64 value = 1;
  uint32 rank = 2;
  uint32 num_workers = 3;
  uint32 root = 4;
}

message BroadcastResponse {
  uint64 value = 1;
}